qiskit/primitives/base_estimator.py

```diff
@@ -0,0 +1,233 @@
+  to be bound to the parameters of the quantum circuits.
+  (list of list of float)
+
+The output is the list of expectation values.
+
+.. math::
+
+    \langle\psi_i(\theta_k)|H_j|\psi_i(\theta_k)\rangle
+
+
+The estimator object is expected to be `close()` d after use or
+accessed inside "with" context
+and the objects are called with parameter values and run options
+(e.g., ``shots`` or number of shots).
+
+Here is an example of how estimator is used.
+
+.. code-block:: python
+
+    from qiskit.circuit.library import RealAmplitudes
+    from qiskit.quantum_info import SparsePauliOp
+
+    psi1 = RealAmplitudes(num_qubits=2, reps=2)
+    psi2 = RealAmplitudes(num_qubits=2, reps=3)
+
+    params1 = psi1.parameters
+    params2 = psi2.parameters
+
+    H1 = SparsePauliOp.from_list([("II", 1), ("IZ", 2), ("XI", 3)])
+    H2 = SparsePauliOp.from_list([("IZ", 1)])
+    H3 = SparsePauliOp.from_list([("ZI", 1), ("ZZ", 1)])
+
+    with Estimator([psi1, psi2], [H1, H2, H3], [params1, params2]) as e:
+        theta1 = [0, 1, 1, 2, 3, 5]
+        theta2 = [0, 1, 1, 2, 3, 5, 8, 13]
+        theta3 = [1, 2, 3, 4, 5, 6]
+
+        # calculate [ <psi1(theta1)|H1|psi1(theta1)> ]
+        psi1_H1_result = e([(0, 0)], [theta1])
+        print(psi1_H1_result)
+
+        # calculate [ <psi1(theta1)|H2|psi1(theta1)>, <psi1(theta1)|H3|psi1(theta1)> ]
+        psi1_H23_result = e([(0, 1), (0, 2)], [theta1]*2)
+        print(psi1_H23_result)
+
+        # calculate [ <psi2(theta2)|H2|psi2(theta2)> ]
+        psi2_H2_result = e([(1, 1)], [theta2])
+        print(psi2_H2_result)
+
+        # calculate [ <psi1(theta1)|H1|psi1(theta1)>, <psi1(theta3)|H1|psi1(theta3)> ]
+        psi1_H1_result2 = e([(0, 0)]*2, [theta1, theta3])
+        print(psi1_H1_result2)
+
+        # calculate [ <psi1(theta1)|H1|psi1(theta1)>,
+        #             <psi2(theta2)|H2|psi2(theta2)>,
+        #             <psi1(theta3)|H3|psi1(theta3)> ]
+        psi12_H23_result = e([(0, 0), (1, 1), (0, 2)], [theta1, theta2, theta3])
+        print(psi12_H23_result)
+"""
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Callable, Iterable, Sequence
+
+from qiskit.circuit import Parameter, QuantumCircuit
+from qiskit.circuit.parametertable import ParameterView
+from qiskit.exceptions import QiskitError
+from qiskit.quantum_info import SparsePauliOp
+
+from .estimator_result import EstimatorResult
+
+
+@dataclass(frozen=True)
+class Group:
+    """The dataclass represents indices of circuit and observable."""
+
+    circuit_index: int
+    observable_index: int
+
+
+class BaseEstimator(ABC):
+    """Estimator base class.
+
+    Base class for Estimator that estimates expectation values of quantum circuits and observables.
+    """
+
+    def __init__(
+        self,
+        circuits: Iterable[QuantumCircuit],
+        observables: Iterable[SparsePauliOp],
+        parameters: Iterable[Iterable[Parameter]] | None = None,
```

Comment:
its ok with `from __future__ import annotations` though, right?

Reply:
I wrote it in the other comment - it won't error in the import, but it'd fail if you tried to type-check using Python 3.7, because it'd error during `typing.get_type_hints`.

Reply:
```
jake@ninetales$ python3.7 -q
>>> from __future__ import annotations
>>> def f(x: tuple | list): pass
...
>>> import typing
>>> typing.get_type_hints(f)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jake/.miniconda3/envs/qiskit37/lib/python3.7/typing.py", line 1013, in get_type_hints
    value = _eval_type(value, globalns, localns)
  File "/Users/jake/.miniconda3/envs/qiskit37/lib/python3.7/typing.py", line 263, in _eval_type
    return t._evaluate(globalns, localns)
  File "/Users/jake/.miniconda3/envs/qiskit37/lib/python3.7/typing.py", line 467, in _evaluate
    eval(self.__forward_code__, globalns, localns),
  File "<string>", line 1, in <module>
TypeError: unsupported operand type(s) for |: 'type' and 'type'
```

Reply:
Oh, the whole `type | type` thing is actually way newer than I thought - in my mind it was ~Python 3.8, but it actually only landed in Python 3.10.

Reply:
If I understand correctly, PEP 604 is from 3.10, but I think it was also available in 3.7 if you used postponed evaluations. But, I prefer Optional to Union[..., None], so I'll fix it here. (I would like to confirm this for future use.)

Reply:
`from __future__ import annotations` will prevent this from raising errors during the import, but type hints are still meant to (eventually) be able to be evaluated as valid Python expressions - the behaviour of `from __future__ import annotations` was defined in Python 3.7 in [PEP 563](https://www.python.org/dev/peps/pep-0563/).  Having the future import will make the code execute correctly, but any library in Python 3.9 or below that tries to evaluate the type hints after the code has been imported will fail. 

Notably, this means that static analysers like `mypy` won't have an issue at all, because they don't actually execute the Python source.  It will (potentially, I've not checked) be an issue for Sphinx, though, which does access the type hints by importing the code and reading the type hints objects out with `typing.get_type_hints`.

Lev and I spoke a bit offline, and it's somewhat a stylistic choice than anything more here.  If it actually causes a problem for Sphinx we should change it to the old style, but otherwise it doesn't matter all that much, because it's not like Terra would pass mypy anyway.

Reply:
When I see PEP 0585 https://www.python.org/dev/peps/pep-0585/#implementation, postponed evaluation allows `tuple[...]` notation. But I'm fine with either. If it passes CI, I think it's good.

Reply:
Oh, I hadn't seen that line in that PEP, thanks.  To be honest, I'm surprised that they merged that into mainline Python, because the "postponed evaluation" PEP includes sections on how to evaluate the type hints correctly at runtime (https://www.python.org/dev/peps/pep-0563/#resolving-type-hints-at-runtime), and the new one makes no mention of how these types would break that.  I suppose probably they were most concerned with static analysis tools, not run-time checks (especially since annotations aren't meant to have any run-time impact).

At any rate, Sphinx doesn't seem to care about this, so from my perspective, absolutely feel free to leave these as-is.

Reply:
Thanks for looking into it jake